perllib/FixMyStreet/App/Controller/Admin/AreaStats.pm

```diff
@@ -0,0 +1,208 @@
+}
+
+sub body_stats : Chained('body_base') : PathPart('') : Args(0) {
+    my ($self, $c) = @_;
+    my $user = $c->user;
+
+    $c->forward('check_user', [$c->stash->{body_id}]);
+    if ($c->get_param('area')) {
+        $c->stash->{area_id} = $c->get_param('area');
+    } else {
+        $c->stash->{area_id} = $c->stash->{body}->id;
+    }
+    $c->forward('stats');
+}
+
+sub ward_stats : Chained('body_base') :PathPart('') : Args(1) {
+    my ($self, $c, $area_id) = @_;
+    my $user = $c->user;
+
+    $c->forward('check_user', [$c->stash->{body_id}, $area_id]);
+
+    $c->stash->{area_id} = $area_id;
+    $c->forward('stats');
+}
+
+sub stats : Private {
+    my ($self, $c) = @_;
+
+    my $date = DateTime->now->subtract(days => 30);
+    my $area_id = $c->stash->{area_id};
+    my $area = mySociety::MaPit::call('area', $area_id );
+    my $user = $c->user;
+
+    $c->forward('/admin/fetch_contacts');
+
+    $c->stash->{template} = 'admin/areastats/area.html';
+    if ($area->{name}) {
+        $c->stash->{area} = $area;
+
+        my $dtf = $c->model('DB')->storage->datetime_parser;
+        my $time = $dtf->format_datetime($date);
+
+        my $params = {
+            'problem.areas' => { like => "%,$area_id,%" },
+            'me.confirmed' => { '>=', $time },
+        };
+
+        my $comments = $c->model('DB::Comment')->search(
+            $params,
+            {
+                group_by => [ 'problem_state' ],
+                select   => [ 'problem_state', { count => 'me.id' } ],
+                as       => [ qw/state state_count/ ],
+                join     => 'problem'
+            }
+        );
+
+        $params = {
+            areas => { like => "%,$area_id,%" },
+        };
+
+        my $problems = $c->model('DB::Problem')->search(
+            $params,
+            {
+                group_by => [ 'category', 'state' ],
+                select   => [ 'category', 'state', { count => 'me.id' } ],
+                as       => [ qw/category state state_count/ ],
+            }
+        );
+
+        my %by_category = map { $_->category => {} } $c->stash->{contacts}->all;
+        my %recent_by_category = map { $_->category => 0 } $c->stash->{contacts}->all;
+
+        my $state_map = {};
+
+        $state_map->{$_} = 'open' foreach FixMyStreet::DB::Result::Problem->open_states;
+        $state_map->{$_} = 'closed' foreach FixMyStreet::DB::Result::Problem->closed_states;
+        $state_map->{$_} = 'fixed' foreach FixMyStreet::DB::Result::Problem->fixed_states;
+        $state_map->{$_} = 'scheduled' foreach ('planned', 'action scheduled');
+
+        for my $p ($problems->all) {
+            my $meta_state = $state_map->{$p->state};
+            $by_category{$p->category}->{$meta_state} += $p->get_column('state_count');
+        }
+
+        $params->{'me.confirmed'} = { '>=', $time };
+
+        my $recent_problems = $c->model('DB::Problem')->search(
```

Comment:
Isn't the result of this query covered by a subset of the results of the last query? As in, with a conditional in the prior loop, you could construct this without going back to the database?

Reply:
It means doing all the summing in code rather than in the DB because it isn't selecting `confirmed` and grouping by that doesn't make sense. And I'm not sure that one loop over every record returned is better than two DB queries, especially doing date maths on every record. :/

Reply:
Fair enough. You could group on the boolean "me.confirmed >= $time" and then you'd have either true or false, but that's probably too confusing. As long as the query speed on the whole dataset is okay, it's fine.